src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,304 @@
+﻿using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+internal partial class SyncGenerator {
+	/// <summary>
+	/// Creates a source file containing generated methods and the necessary
+	/// namespaces, using directives, and type declarations for them to live in
+	/// (but only the necessary ones).
+	/// </summary>
+	internal sealed class FileCollector {
+		private static readonly SyntaxToken NoneToken
+			= SyntaxFactory.Token( SyntaxKind.None );
+
+		private static readonly SyntaxTriviaList EmptyTrivia
+			= SyntaxFactory.TriviaList();
+
+		private readonly CompilationUnitSyntax m_root;
+
+		// This is a mutable dictionary because we will remove from it as we go
+		// and check that it is empty at the end, just to be safe.
+		private readonly Dictionary<SyntaxNode, IEnumerable<string>> m_methods;
+
+		/// <summary>
+		/// Wraps the sort of syntax that we will append as part of a preamble, to
+		/// defer calls to ToFullString()
+		/// </summary>
+		private readonly struct PieceOfSyntax {
+			public object Data { get; init; }
+
+			public static implicit operator PieceOfSyntax( SyntaxNode node )
+				=> new() { Data = node };
+
+			public static implicit operator PieceOfSyntax( SyntaxToken token )
+				=> new() { Data = token };
+
+			public static implicit operator PieceOfSyntax( SyntaxList<UsingDirectiveSyntax> list )
+				=> new() { Data = list };
+
+			public override string ToString() => Data switch {
+				SyntaxNode n => n.ToFullString(),
+				SyntaxToken t => t.ToFullString(),
+				SyntaxList<UsingDirectiveSyntax> nl => nl.ToFullString(),
+				_ => throw new BugException( "invalid piece of syntax " + Data.GetType().Name )
+			};
+		}
+
+		// When we recurse into a namespace or type, we're not sure if the
+		// preamble ("namespace foo {", "\t\tpartial class Bar<T> {", etc.) is
+		// necessary, because maybe there are no generated methods.
+		// Using a LinkedList<string> as a double-ended queue.
+		private readonly LinkedList<PieceOfSyntax[]> m_preambles = new();
+
+		private readonly StringBuilder m_out = new();
+
+		private FileCollector(
+			CompilationUnitSyntax root,
+			Dictionary<SyntaxNode, IEnumerable<string>> methods
+		) {
+			m_root = root;
+			m_methods = methods;
+		}
+
+		public static FileCollector Create(
+			CompilationUnitSyntax root,
+			ImmutableArray<(SyntaxNode Parent, string Source)> methods

```

Comment:
Parent is _usually_ a type?

Reply:
Or namespace or compilation unit

Reply:
Parent of methods? Nah. WriteMethods implies it's just TypeDeclarationSyntax

Reply:
It looks like the bare ones aren't directly under `CompilationUnitSyntax` but instead `GlobalStatementSyntax`: https://sharplab.io/#v2:C4LghgzsA+CwBQABALAAgGIHtMAoCUqA3ggL5A==

Reply:
(not saying this is something that's going to work or that we want to support it (probably not? weird with partials))

Reply:
I did what you were getting at here and changed the interface to assume that parents are `TypeDeclarationSyntax`. It will mean some more casting/error handling in the upper layer but that is fine. It was going to crash somewhere ultimately, moving it earlier will mean that we separate "you have a method in an unsupported kind of type" from the error that will happen for bare methods/whatever crazy new syntax they come up with next... seems good to me.